wiring/src/spark_wiring_ledger.cpp

```diff
@@ -0,0 +1,434 @@
+            // Treat empty data as an empty map
+            data = LedgerData();
+            return 0;
+        }
+        LOG(ERROR, "Failed to decode ledger data: %d", r);
+        return r;
+    }
+    if (!v.isMap()) {
+        LOG(ERROR, "Unexpected type of ledger data");
+        return Error::BAD_DATA;
+    }
+    data = std::move(v);
+    return 0;
+}
+
+} // namespace
+
+int Ledger::set(const LedgerData& data, SetMode mode) {
+    if (!isValid()) {
+        return Error::INVALID_STATE;
+    }
+    if (mode == Ledger::REPLACE) {
+        CHECK(setLedgerData(instance_, data));
+    } else {
+        LedgerData d;
+        CHECK(getLedgerData(instance_, d));
+        for (auto& e: data.variantMap()) {
+            if (!d.set(e.first, e.second)) {
+                return Error::NO_MEMORY;
+            }
+        }
+        CHECK(setLedgerData(instance_, d));
+    }
+    return 0;
+}
+
+LedgerData Ledger::get() const {
+    if (!isValid()) {
+        return LedgerData();
+    }
+    LedgerData data;
+    if (getLedgerData(instance_, data) < 0) {
+        return LedgerData();
+    }
+    return data;
+}
+
+int64_t Ledger::lastUpdated() const {
+    ledger_info info = {};
+    if (!isValid() || getLedgerInfo(instance_, info) < 0) {
+        return 0;
+    }
+    return info.last_updated;
+}
+
+int64_t Ledger::lastSynced() const {
+    ledger_info info = {};
+    if (!isValid() || getLedgerInfo(instance_, info) < 0) {
+        return 0;
+    }
+    return info.last_synced;
+}
+
+size_t Ledger::dataSize() const {
+    ledger_info info = {};
+    if (!isValid() || getLedgerInfo(instance_, info) < 0) {
+        return 0;
+    }
+    return info.data_size;
+}
+
+const char* Ledger::name() const {
+    ledger_info info = {};
+    if (!isValid() || getLedgerInfo(instance_, info) < 0) {
+        return "";
+    }
+    return info.name;
+}
+
+LedgerScope Ledger::scope() const {
+    ledger_info info = {};
+    if (!isValid() || getLedgerInfo(instance_, info) < 0) {
+        return LedgerScope::UNKNOWN;
+    }
+    return static_cast<LedgerScope>(info.scope);
+}
+
+bool Ledger::isWritable() const {
+    ledger_info info = {};
+    if (!isValid() || getLedgerInfo(instance_, info) < 0) {
+        return false;
+    }
+    // It's allowed to write to a ledger while its sync direction is unknown
+    return info.sync_direction == LEDGER_SYNC_DIRECTION_DEVICE_TO_CLOUD ||
+            info.sync_direction == LEDGER_SYNC_DIRECTION_UNKNOWN;
+}
+
+int Ledger::onSync(OnSyncCallback callback, void* arg) {
+    if (!isValid()) {
+        return Error::INVALID_STATE;
+    }
+    if (!callback) {
+        return onSync(Ledger::OnSyncFunction());
+    }
+    return onSync([callback, arg](Ledger ledger) {
+        callback(std::move(ledger), arg);
+    });
+}
+
+int Ledger::onSync(OnSyncFunction callback) {
+    if (!isValid()) {
+        return Error::INVALID_STATE;
+    }
+    return setSyncCallback(instance_, std::move(callback));
+}
```

Comment:
On the topic of making things consistent, would we want to make all the methods to provide callbacks in the user application consistent using the pattern above? We have several patterns in the Wiring layer for callbacks, like Particle.subscribe and System.on. Could we mark some of the patterns like passing a pointer to a method and a pointer to an instance deprecated and tell developers to use lambdas? We can update the public docs as well.

Reply:
I have a minor concern related to the fact that `std::function` is not exception-safe so, strictly speaking, it should be avoided in embedded code, but otherwise I'm not opposed to the suggested change in our approach. It would also be nice to have a discussion about the default OOM handling behavior in general as part of that decision.

Reply:
Btw, I simplified the code that registers the `onSync` callback following your feedback in one the previous PRs. Hope it's clearer now.

Reply:
Yes, I like the simplication. Thank you.

Reply:
@avtolstoy This can be a topic to discuss with a broader group of Device OS engineers.

Reply:
:+1: But just to say in advance this is the same pattern that we use in recently introduced APIs. GuoHui and I have gone through this during BLE introduction in detail and have come to the conclusion that function pointer + std::function wrapper most of the times is sufficient. One extra step is:
```c++
template<typename T>
void onBlah(void(T::*callback)(void arg1, void arg2), T* instance);
```